File: src/main/kotlin/leetCode/easy/LargestPerimeterTriangle.kt
package leetCode.easy

import java.util.Collections
import java.util.PriorityQueue

class LargestPerimeterTriangle {
    fun largestPerimeter(nums: IntArray): Int {
        val maxHeap = PriorityQueue<Int>(Collections.reverseOrder()).also {
            it.addAll(nums.toList())
        }

        while (maxHeap.size >= 3) {
            val max = maxHeap.poll()
            val secondMax = maxHeap.poll()
            val thirdMax = maxHeap.peek()

            if (secondMax + thirdMax > max) {
                return secondMax + thirdMax + max
            } else {
                maxHeap.add(secondMax)
            }
        }

        return 0
    }
}

File: go/leetCode/easy/LargestPerimeterTriangle.go
package easy

import "sort"

func largestPerimeter(nums []int) int {
	sort.Ints(nums)

	for i := len(nums) - 1; i >= 2; i-- {
		max := nums[i]
		secondMax := nums[i-1]
		thirdMax := nums[i-2]

		if secondMax+thirdMax > max {
			return max + secondMax + thirdMax
		}
	}

	return 0
}
